File: src/google_slides_mcp/tools/creation.py
"""Creation tools for adding elements to Google Slides.

Tools for creating slides, text boxes, images, and shapes with
intuitive parameters using inches instead of EMUs.
"""

from typing import TYPE_CHECKING, Literal
import uuid

if TYPE_CHECKING:
    from fastmcp import Context, FastMCP


# Predefined slide layouts
LAYOUT_TYPES = Literal[
    "BLANK",
    "TITLE",
    "TITLE_AND_BODY",
    "TITLE_AND_TWO_COLUMNS",
    "TITLE_ONLY",
    "SECTION_HEADER",
    "ONE_COLUMN_TEXT",
    "MAIN_POINT",
    "BIG_NUMBER",
    "CAPTION_ONLY",
]


def register_creation_tools(mcp: "FastMCP") -> None:
    """Register creation tools with the MCP application.

    Args:
        mcp: The FastMCP application instance
    """

    @mcp.tool()
    async def create_slide(
        presentation_id: str,
        layout: LAYOUT_TYPES = "BLANK",
        insertion_index: int | None = None,
        ctx: "Context",
    ) -> dict:
        """Create a new slide with the specified layout.

        Args:
            presentation_id: The presentation to add the slide to
            layout: The layout type for the new slide
            insertion_index: Position to insert (None = append at end)

        Returns:
            Dictionary with:
            - slide_id: ID of the new slide
            - placeholder_ids: Mapping of placeholder types to their IDs
        """
        from google_slides_mcp.auth.middleware import GoogleAuthMiddleware
        from google_slides_mcp.services.slides_service import SlidesService

        middleware = GoogleAuthMiddleware()
        credentials = await middleware.extract_credentials(ctx)
        service = SlidesService(credentials)

        # Get presentation to find the layout
        presentation = await service.get_presentation(presentation_id)

        # Find matching layout
        layout_id = None
        for layout_obj in presentation.get("layouts", []):
            layout_props = layout_obj.get("layoutProperties", {})
            if layout_props.get("name", "").upper() == layout:
                layout_id = layout_obj.get("objectId")
                break

        # Generate a unique ID for the new slide
        slide_id = f"slide_{uuid.uuid4().hex[:8]}"

        request: dict = {
            "createSlide": {
                "objectId": slide_id,
                "slideLayoutReference": {"predefinedLayout": layout},
            }
        }

        if layout_id:
            request["createSlide"]["slideLayoutReference"] = {"layoutId": layout_id}

        if insertion_index is not None:
            request["createSlide"]["insertionIndex"] = insertion_index

        await service.batch_update(presentation_id, [request])

        # Get the created slide to find placeholder IDs
        slide_info = await service.get_page(presentation_id, slide_id)
        placeholder_ids = {}
        for element in slide_info.get("pageElements", []):
            placeholder = element.get("shape", {}).get("placeholder", {})
            if placeholder:
                placeholder_type = placeholder.get("type", "UNKNOWN")
                placeholder_ids[placeholder_type] = element.get("objectId")

        return {
            "slide_id": slide_id,
            "placeholder_ids": placeholder_ids,
        }

    @mcp.tool()
    async def add_text_box(
        presentation_id: str,
        slide_id: str,
        text: str,
        x: float = 1.0,
        y: float = 1.0,
        width: float = 4.0,
        height: float = 1.0,
        font_size: int = 18,
        font_family: str = "Arial",
        bold: bool = False,
        italic: bool = False,
        color: str = "#000000",
        alignment: Literal["LEFT", "CENTER", "RIGHT"] = "LEFT",
        ctx: "Context",
    ) -> dict:
        """Add a text box with styling to a slide.

        Args:
            presentation_id: The presentation ID
            slide_id: The slide to add the text box to
            text: The text content
            x: X position in inches from left edge
            y: Y position in inches from top edge
            width: Width in inches
            height: Height in inches
            font_size: Font size in points
            font_family: Font family name
            bold: Whether text is bold
            italic: Whether text is italic
            color: Text color as hex (e.g., "#FF0000")
            alignment: Text alignment within the box

        Returns:
            Dictionary with the created element ID
        """
        from google_slides_mcp.auth.middleware import GoogleAuthMiddleware
        from google_slides_mcp.services.slides_service import SlidesService
        from google_slides_mcp.utils.colors import hex_to_rgb
        from google_slides_mcp.utils.units import inches_to_emu, points_to_emu

        middleware = GoogleAuthMiddleware()
        credentials = await middleware.extract_credentials(ctx)
        service = SlidesService(credentials)

        # Generate unique ID
        element_id = f"textbox_{uuid.uuid4().hex[:8]}"

        requests = [
            # Create the text box shape
            {
                "createShape": {
                    "objectId": element_id,
                    "shapeType": "TEXT_BOX",
                    "elementProperties": {
                        "pageObjectId": slide_id,
                        "size": {
                            "width": {"magnitude": inches_to_emu(width), "unit": "EMU"},
                            "height": {"magnitude": inches_to_emu(height), "unit": "EMU"},
                        },
                        "transform": {
                            "scaleX": 1,
                            "scaleY": 1,
                            "shearX": 0,
                            "shearY": 0,
                            "translateX": inches_to_emu(x),
                            "translateY": inches_to_emu(y),
                            "unit": "EMU",
                        },
                    },
                }
            },
            # Insert the text
            {
                "insertText": {
                    "objectId": element_id,
                    "text": text,
                    "insertionIndex": 0,
                }
            },
            # Style the text
            {
                "updateTextStyle": {
                    "objectId": element_id,
                    "style": {
                        "fontFamily": font_family,
                        "fontSize": {"magnitude": font_size, "unit": "PT"},
                        "bold": bold,
                        "italic": italic,
                        "foregroundColor": {
                            "opaqueColor": {"rgbColor": hex_to_rgb(color)}
                        },
                    },
                    "fields": "fontFamily,fontSize,bold,italic,foregroundColor",
                    "textRange": {"type": "ALL"},
                }
            },
            # Set paragraph alignment
            {
                "updateParagraphStyle": {
                    "objectId": element_id,
                    "style": {"alignment": alignment},
                    "fields": "alignment",
                    "textRange": {"type": "ALL"},
                }
            },
        ]

        await service.batch_update(presentation_id, requests)

        return {"element_id": element_id}

    @mcp.tool()
    async def add_image(
        presentation_id: str,
        slide_id: str,
        image_url: str,
        x: float | None = None,
        y: float | None = None,
        width: float | None = None,
        height: float | None = None,
        horizontal_align: Literal["left", "center", "right"] | None = None,
        vertical_align: Literal["top", "center", "bottom"] | None = None,
        ctx: "Context",
    ) -> dict:
        """Add an image to a slide from a URL.

        If only width OR height is specified, aspect ratio is preserved.
        Use alignment for quick positioning without calculating coordinates.

        Args:
            presentation_id: The presentation ID
            slide_id: The slide to add the image to
            image_url: URL of the image (must be publicly accessible)
            x: X position in inches (optional if using alignment)
            y: Y position in inches (optional if using alignment)
            width: Width in inches (optional, preserves aspect ratio)
            height: Height in inches (optional, preserves aspect ratio)
            horizontal_align: Horizontal alignment on slide
            vertical_align: Vertical alignment on slide

        Returns:
            Dictionary with the created element ID
        """
        from google_slides_mcp.auth.middleware import GoogleAuthMiddleware
        from google_slides_mcp.services.slides_service import SlidesService
        from google_slides_mcp.utils.transforms import (
            SLIDE_SIZES,
            calculate_alignment_position,
        )
        from google_slides_mcp.utils.units import inches_to_emu

        middleware = GoogleAuthMiddleware()
        credentials = await middleware.extract_credentials(ctx)
        service = SlidesService(credentials)

        # Generate unique ID
        element_id = f"image_{uuid.uuid4().hex[:8]}"

        # Default size if not specified
        default_width = 4.0
        default_height = 3.0

        img_width = width if width is not None else default_width
        img_height = height if height is not None else default_height

        # Calculate position
        slide_size = SLIDE_SIZES["16:9"]

        if horizontal_align is not None or vertical_align is not None:
            pos_x, pos_y = calculate_alignment_position(
                slide_size,
                inches_to_emu(img_width),
                inches_to_emu(img_height),
                horizontal=horizontal_align,
                vertical=vertical_align,
            )
        else:
            pos_x = inches_to_emu(x if x is not None else 1.0)
            pos_y = inches_to_emu(y if y is not None else 1.0)

        request = {
            "createImage": {
                "objectId": element_id,
                "url": image_url,
                "elementProperties": {
                    "pageObjectId": slide_id,
                    "size": {
                        "width": {"magnitude": inches_to_emu(img_width), "unit": "EMU"},
                        "height": {"magnitude": inches_to_emu(img_height), "unit": "EMU"},
                    },
                    "transform": {
                        "scaleX": 1,
                        "scaleY": 1,
                        "shearX": 0,
                        "shearY": 0,
                        "translateX": pos_x,
                        "translateY": pos_y,
                        "unit": "EMU",
                    },
                },
            }
        }

        await service.batch_update(presentation_id, [request])

        return {"element_id": element_id}

    @mcp.tool()
    async def add_shape(
        presentation_id: str,
        slide_id: str,
        shape_type: str,
        x: float,
        y: float,
        width: float,
        height: float,
        fill_color: str | None = None,
        outline_color: str | None = "#000000",
        outline_weight: float = 1.0,
        ctx: "Context",
    ) -> dict:
        """Add a shape to a slide.

        Args:
            presentation_id: The presentation ID
            slide_id: The slide to add the shape to
            shape_type: Shape type (RECTANGLE, ELLIPSE, TRIANGLE, STAR_5, etc.)
                See Google Slides API ShapeType enum for full list.
            x: X position in inches
            y: Y position in inches
            width: Width in inches
            height: Height in inches
            fill_color: Fill color as hex (None for no fill)
            outline_color: Outline color as hex (None for no outline)
            outline_weight: Outline weight in points

        Returns:
            Dictionary with the created element ID
        """
        from google_slides_mcp.auth.middleware import GoogleAuthMiddleware
        from google_slides_mcp.services.slides_service import SlidesService
        from google_slides_mcp.utils.colors import hex_to_rgb
        from google_slides_mcp.utils.units import inches_to_emu, points_to_emu

        middleware = GoogleAuthMiddleware()
        credentials = await middleware.extract_credentials(ctx)
        service = SlidesService(credentials)

        # Generate unique ID
        element_id = f"shape_{uuid.uuid4().hex[:8]}"

        requests = [
            {
                "createShape": {
                    "objectId": element_id,
                    "shapeType": shape_type,
                    "elementProperties": {
                        "pageObjectId": slide_id,
                        "size": {
                            "width": {"magnitude": inches_to_emu(width), "unit": "EMU"},
                            "height": {"magnitude": inches_to_emu(height), "unit": "EMU"},
                        },
                        "transform": {
                            "scaleX": 1,
                            "scaleY": 1,
                            "shearX": 0,
                            "shearY": 0,
                            "translateX": inches_to_emu(x),
                            "translateY": inches_to_emu(y),
                            "unit": "EMU",
                        },
                    },
                }
            }
        ]

        # Build shape properties update
        shape_props: dict = {}
        fields = []

        if fill_color is not None:
            shape_props["shapeBackgroundFill"] = {
                "solidFill": {"color": {"rgbColor": hex_to_rgb(fill_color)}}
            }
            fields.append("shapeBackgroundFill")

        if outline_color is not None:
            shape_props["outline"] = {
                "outlineFill": {
                    "solidFill": {"color": {"rgbColor": hex_to_rgb(outline_color)}}
                },
                "weight": {"magnitude": outline_weight, "unit": "PT"},
            }
            fields.append("outline")

        if fields:
            requests.append(
                {
                    "updateShapeProperties": {
                        "objectId": element_id,
                        "shapeProperties": shape_props,
                        "fields": ",".join(fields),
                    }
                }
            )

        await service.batch_update(presentation_id, requests)

        return {"element_id": element_id}

File: packages/cloudflare/src/tools/creation.ts
/**
 * Creation tools for adding elements to Google Slides.
 *
 * Tools for creating slides, text boxes, images, and shapes with
 * intuitive parameters using inches instead of EMUs.
 */

import type { McpServer } from "@modelcontextprotocol/sdk/server/mcp.js";
import { z } from "zod";
import { SlidesClient } from "../api/slides-client.js";
import type { TokenManager } from "../api/token-manager.js";
import { inchesToEmu, emuToInches } from "../utils/units.js";
import { hexToRgb } from "../utils/colors.js";
import { SLIDE_SIZES, calculateAlignmentPosition, extractElementBounds } from "../utils/transforms.js";

/** Map user-facing alignment values to Google Slides API ParagraphStyle.Alignment enum. */
function toApiAlignment(alignment: string): string {
  switch (alignment) {
    case "LEFT": return "START";
    case "RIGHT": return "END";
    default: return alignment;
  }
}

/**
 * Unescape literal \n and \t sequences that LLMs sometimes double-escape
 * in tool call parameters.
 */
function unescapeText(text: string): string {
  return text.replace(/\\n/g, "\n").replace(/\\t/g, "\t");
}

function generateId(prefix: string): string {
  return `${prefix}_${Math.random().toString(36).substring(2, 10)}`;
}

/**
 * Register creation tools with the MCP server.
 */
export function registerCreationTools(
  server: McpServer,
  tokenManager: TokenManager
): void {
  const client = new SlidesClient(tokenManager);

  /**
   * create_slide - Create a new slide with the specified layout
   */
  const PREDEFINED_LAYOUTS = new Set([
    "BLANK", "TITLE", "TITLE_AND_BODY", "TITLE_AND_TWO_COLUMNS",
    "TITLE_ONLY", "SECTION_HEADER", "ONE_COLUMN_TEXT", "MAIN_POINT",
    "BIG_NUMBER", "CAPTION_ONLY",
  ]);

  server.tool(
    "create_slide",
    `Create a new slide with the specified layout. Returns placeholder positions and sizes in inches for spatial awareness when placing additional elements. Accepts either a predefined layout name (BLANK, TITLE, TITLE_AND_BODY, TITLE_AND_TWO_COLUMNS, TITLE_ONLY, SECTION_HEADER, ONE_COLUMN_TEXT, MAIN_POINT, BIG_NUMBER, CAPTION_ONLY) or a custom layout object ID from the presentation (use list_layouts to discover available layouts).`,
    {
      presentation_id: z.string().describe("The presentation to add the slide to"),
      layout: z.string().default("BLANK").describe("Predefined layout name or custom layout object ID"),
      insertion_index: z.number().optional().describe("Position to insert (null = append at end)"),
    },
    async ({ presentation_id, layout, insertion_index }) => {
      try {
        const slideId = generateId("slide");

        const slideLayoutReference = PREDEFINED_LAYOUTS.has(layout)
          ? { predefinedLayout: layout }
          : { layoutId: layout };

        const request: Record<string, unknown> = {
          createSlide: {
            objectId: slideId,
            slideLayoutReference,
          },
        };

        if (insertion_index !== undefined) {
          (request.createSlide as Record<string, unknown>).insertionIndex = insertion_index;
        }

        await client.batchUpdate(presentation_id, [request]);

        // Get the created slide to find placeholders with their bounds
        const slideInfo = await client.getPage(presentation_id, slideId);
        const placeholders: Array<{
          type: string;
          id: string;
          x: number;
          y: number;
          width: number;
          height: number;
        }> = [];

        for (const element of slideInfo.pageElements ?? []) {
          const placeholder = element.shape?.placeholder;
          if (placeholder) {
            const entry: {
              type: string;
              id: string;
              x: number;
              y: number;
              width: number;
              height: number;
            } = {
              type: placeholder.type ?? "UNKNOWN",
              id: element.objectId ?? "",
              x: 0,
              y: 0,
              width: 0,
              height: 0,
            };

            if (element.size && element.transform) {
              const bounds = extractElementBounds(element);
              entry.x = Math.round(emuToInches(bounds.x) * 100) / 100;
              entry.y = Math.round(emuToInches(bounds.y) * 100) / 100;
              entry.width = Math.round(emuToInches(bounds.width) * 100) / 100;
              entry.height = Math.round(emuToInches(bounds.height) * 100) / 100;
            }

            placeholders.push(entry);
          }
        }

        return {
          content: [
            {
              type: "text" as const,
              text: JSON.stringify({
                slide_id: slideId,
                placeholders,
              }, null, 2),
            },
          ],
        };
      } catch (error) {
        return {
          content: [
            {
              type: "text" as const,
              text: `Error: ${error instanceof Error ? error.message : String(error)}`,
            },
          ],
          isError: true,
        };
      }
    }
  );

  /**
   * add_text_box - Add a text box with styling to a slide
   */
  server.tool(
    "add_text_box",
    "Add a text box with styling to a slide.",
    {
      presentation_id: z.string().describe("The presentation ID"),
      slide_id: z.string().describe("The slide to add the text box to"),
      text: z.string().describe("The text content"),
      x: z.number().default(1).describe("X position in inches from left edge"),
      y: z.number().default(1).describe("Y position in inches from top edge"),
      width: z.number().default(4).describe("Width in inches"),
      height: z.number().default(1).describe("Height in inches"),
      font_size: z.number().default(18).describe("Font size in points"),
      font_family: z.string().default("Arial").describe("Font family name"),
      bold: z.boolean().default(false).describe("Whether text is bold"),
      italic: z.boolean().default(false).describe("Whether text is italic"),
      color: z.string().default("#000000").describe("Text color as hex"),
      alignment: z.enum(["LEFT", "CENTER", "RIGHT"]).default("LEFT").describe("Text alignment"),
      autofit: z.enum(["none", "shrink_text", "resize_shape"]).default("none").describe("Autofit behavior: none, shrink_text (shrink text to fit), or resize_shape (grow shape to fit text)"),
    },
    async ({ presentation_id, slide_id, text, x, y, width, height, font_size, font_family, bold, italic, color, alignment, autofit }) => {
      try {
        const elementId = generateId("textbox");

        const requests: Record<string, unknown>[] = [
          // Create the text box shape
          {
            createShape: {
              objectId: elementId,
              shapeType: "TEXT_BOX",
              elementProperties: {
                pageObjectId: slide_id,
                size: {
                  width: { magnitude: inchesToEmu(width), unit: "EMU" },
                  height: { magnitude: inchesToEmu(height), unit: "EMU" },
                },
                transform: {
                  scaleX: 1,
                  scaleY: 1,
                  shearX: 0,
                  shearY: 0,
                  translateX: inchesToEmu(x),
                  translateY: inchesToEmu(y),
                  unit: "EMU",
                },
              },
            },
          },
          // Insert the text
          {
            insertText: {
              objectId: elementId,
              text: unescapeText(text),
              insertionIndex: 0,
            },
          },
          // Style the text
          {
            updateTextStyle: {
              objectId: elementId,
              style: {
                fontFamily: font_family,
                fontSize: { magnitude: font_size, unit: "PT" },
                bold,
                italic,
                foregroundColor: {
                  opaqueColor: { rgbColor: hexToRgb(color) },
                },
              },
              fields: "fontFamily,fontSize,bold,italic,foregroundColor",
              textRange: { type: "ALL" },
            },
          },
          // Set paragraph alignment
          {
            updateParagraphStyle: {
              objectId: elementId,
              style: { alignment: toApiAlignment(alignment) },
              fields: "alignment",
              textRange: { type: "ALL" },
            },
          },
        ];

        // Apply autofit if requested
        if (autofit !== "none") {
          requests.push({
            updateShapeProperties: {
              objectId: elementId,
              shapeProperties: {
                autofit: {
                  autofitType: autofit === "shrink_text" ? "TEXT_AUTOFIT" : "SHAPE_AUTOFIT",
                },
              },
              fields: "autofit.autofitType",
            },
          });
        }

        await client.batchUpdate(presentation_id, requests);

        return {
          content: [
            {
              type: "text" as const,
              text: JSON.stringify({ element_id: elementId }, null, 2),
            },
          ],
        };
      } catch (error) {
        return {
          content: [
            {
              type: "text" as const,
              text: `Error: ${error instanceof Error ? error.message : String(error)}`,
            },
          ],
          isError: true,
        };
      }
    }
  );

  /**
   * add_image - Add an image to a slide from a URL
   */
  server.tool(
    "add_image",
    "Add an image to a slide from a URL. If only width OR height is specified, aspect ratio is preserved. Use alignment for quick positioning.",
    {
      presentation_id: z.string().describe("The presentation ID"),
      slide_id: z.string().describe("The slide to add the image to"),
      image_url: z.string().describe("URL of the image (must be publicly accessible)"),
      x: z.number().optional().describe("X position in inches"),
      y: z.number().optional().describe("Y position in inches"),
      width: z.number().optional().describe("Width in inches"),
      height: z.number().optional().describe("Height in inches"),
      horizontal_align: z.enum(["left", "center", "right"]).optional().describe("Horizontal alignment on slide"),
      vertical_align: z.enum(["top", "center", "bottom"]).optional().describe("Vertical alignment on slide"),
    },
    async ({ presentation_id, slide_id, image_url, x, y, width, height, horizontal_align, vertical_align }) => {
      try {
        const elementId = generateId("image");

        // Default size if not specified
        const imgWidth = width ?? 4.0;
        const imgHeight = height ?? 3.0;

        // Calculate position
        const slideSize = SLIDE_SIZES["16:9"];
        let posX: number;
        let posY: number;

        if (horizontal_align !== undefined || vertical_align !== undefined) {
          const [alignX, alignY] = calculateAlignmentPosition(
            slideSize,
            inchesToEmu(imgWidth),
            inchesToEmu(imgHeight),
            horizontal_align,
            vertical_align
          );
          posX = alignX;
          posY = alignY;
        } else {
          posX = inchesToEmu(x ?? 1.0);
          posY = inchesToEmu(y ?? 1.0);
        }

        const request = {
          createImage: {
            objectId: elementId,
            url: image_url,
            elementProperties: {
              pageObjectId: slide_id,
              size: {
                width: { magnitude: inchesToEmu(imgWidth), unit: "EMU" },
                height: { magnitude: inchesToEmu(imgHeight), unit: "EMU" },
              },
              transform: {
                scaleX: 1,
                scaleY: 1,
                shearX: 0,
                shearY: 0,
                translateX: posX,
                translateY: posY,
                unit: "EMU",
              },
            },
          },
        };

        await client.batchUpdate(presentation_id, [request]);

        return {
          content: [
            {
              type: "text" as const,
              text: JSON.stringify({ element_id: elementId }, null, 2),
            },
          ],
        };
      } catch (error) {
        return {
          content: [
            {
              type: "text" as const,
              text: `Error: ${error instanceof Error ? error.message : String(error)}`,
            },
          ],
          isError: true,
        };
      }
    }
  );

  /**
   * add_shape - Add a shape to a slide
   */
  server.tool(
    "add_shape",
    "Add a shape to a slide.",
    {
      presentation_id: z.string().describe("The presentation ID"),
      slide_id: z.string().describe("The slide to add the shape to"),
      shape_type: z.string().describe("Shape type (RECTANGLE, ELLIPSE, TRIANGLE, STAR_5, etc.)"),
      x: z.number().describe("X position in inches"),
      y: z.number().describe("Y position in inches"),
      width: z.number().describe("Width in inches"),
      height: z.number().describe("Height in inches"),
      fill_color: z.string().optional().describe("Fill color as hex (null for no fill)"),
      outline_color: z.string().default("#000000").optional().describe("Outline color as hex"),
      outline_weight: z.number().default(1).describe("Outline weight in points"),
    },
    async ({ presentation_id, slide_id, shape_type, x, y, width, height, fill_color, outline_color, outline_weight }) => {
      try {
        const elementId = generateId("shape");

        const requests: Record<string, unknown>[] = [
          {
            createShape: {
              objectId: elementId,
              shapeType: shape_type,
              elementProperties: {
                pageObjectId: slide_id,
                size: {
                  width: { magnitude: inchesToEmu(width), unit: "EMU" },
                  height: { magnitude: inchesToEmu(height), unit: "EMU" },
                },
                transform: {
                  scaleX: 1,
                  scaleY: 1,
                  shearX: 0,
                  shearY: 0,
                  translateX: inchesToEmu(x),
                  translateY: inchesToEmu(y),
                  unit: "EMU",
                },
              },
            },
          },
        ];

        // Build shape properties update
        const shapeProps: Record<string, unknown> = {};
        const fields: string[] = [];

        if (fill_color) {
          shapeProps.shapeBackgroundFill = {
            solidFill: { color: { rgbColor: hexToRgb(fill_color) } },
          };
          fields.push("shapeBackgroundFill");
        }

        if (outline_weight <= 0) {
          // Zero or negative weight: suppress outline entirely
          shapeProps.outline = { propertyState: "NOT_RENDERED" };
          fields.push("outline");
        } else if (outline_color) {
          shapeProps.outline = {
            outlineFill: {
              solidFill: { color: { rgbColor: hexToRgb(outline_color) } },
            },
            weight: { magnitude: outline_weight, unit: "PT" },
          };
          fields.push("outline");
        }

        if (fields.length > 0) {
          requests.push({
            updateShapeProperties: {
              objectId: elementId,
              shapeProperties: shapeProps,
              fields: fields.join(","),
            },
          });
        }

        await client.batchUpdate(presentation_id, requests);

        return {
          content: [
            {
              type: "text" as const,
              text: JSON.stringify({ element_id: elementId }, null, 2),
            },
          ],
        };
      } catch (error) {
        return {
          content: [
            {
              type: "text" as const,
              text: `Error: ${error instanceof Error ? error.message : String(error)}`,
            },
          ],
          isError: true,
        };
      }
    }
  );
}
